src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php

```diff
@@ -309,7 +309,12 @@ private function hasUserChanged(UserInterface $user): bool
         $userIdentifier = function ($user) {
             return method_exists($user, 'getUserIdentifier') ? $user->getUserIdentifier() : $user->getUsername();
         };
-        if ($userIdentifier($this->user) !== $userIdentifier($user)) {
+        // if the username/identifier is an object with an eqials-method (like Uuid or Ulid), prefer that for a comparison
+        if (\is_object($userIdentifier($this->user)) && method_exists($userIdentifier($this->user), 'equals')) {
```

Comment:
I don't feel comfortable with using duck typing here. Maybe we can check for `instanceof AbstractUid` instead?

Reply:
what about an even more generic interface as a contract?
`EquatableInterface::equals(mixed): bool`

Reply:
I did think about that, but then decided not to implement a lex Uid, but a more abstract solution that solves the problem for all object-based identifiers that have an equals method. Maybe someone uses ShortID or other implementations of Uid.

But you are right, that duck typing opens the door widely for nasty other bugs, so maybe it 's not worth it and a lex Uid is a more stable solution. I'll change that.